plugins/producer_plugin/include/eosio/producer_plugin/block_timing_util.hpp

```diff
@@ -25,22 +52,95 @@ namespace block_timing_util {
+         // production round. In this case, we just use the hard deadline.
+         const auto hard_deadline = block_time.to_time_point() - fc::microseconds(chain::config::block_interval_us - cpu_effort_us);
+         return std::min(hard_deadline, now + fc::microseconds(cpu_effort_us));
+      }
+   }
+
+   inline uint32_t calculate_next_block_slot(const chain::account_name& producer_name, uint32_t current_block_slot, uint32_t block_num,
+                                             const std::vector<chain::producer_authority>& active_schedule, const producer_watermarks& prod_watermarks) {
+      // determine if this producer is in the active schedule and if so, where
+      auto itr =
+         std::find_if(active_schedule.begin(), active_schedule.end(), [&](const auto& asp) { return asp.producer_name == producer_name; });
+      if (itr == active_schedule.end()) {
+         // this producer is not in the active producer set
+         return UINT32_MAX;
+      }
+
+      size_t   producer_index = itr - active_schedule.begin();
+      uint32_t minimum_offset = 1; // must at least be the "next" block
+
+      // account for a watermark in the future which is disqualifying this producer for now
+      // this is conservative assuming no blocks are dropped.  If blocks are dropped the watermark will
+      // disqualify this producer for longer but it is assumed they will wake up, determine that they
+      // are disqualified for longer due to skipped blocks and re-calculate their next block with better
+      // information then
+      auto current_watermark = prod_watermarks.get_watermark(producer_name);
+      if (current_watermark) {
+         const auto watermark = *current_watermark;
+         if (watermark.first > block_num) {
+            // if I have a watermark block number then I need to wait until after that watermark
+            minimum_offset = watermark.first - block_num + 1;
          }
+         if (watermark.second.slot > current_block_slot) {
+            // if I have a watermark block timestamp then I need to wait until after that watermark timestamp
+            minimum_offset = std::max(minimum_offset, watermark.second.slot - current_block_slot + 1);
+         }
+      }
+
+      // this producers next opportunity to produce is the next time its slot arrives after or at the calculated minimum
+      uint32_t minimum_slot = current_block_slot + minimum_offset;
+      size_t   minimum_slot_producer_index =
+         (minimum_slot % (active_schedule.size() * chain::config::producer_repetitions)) / chain::config::producer_repetitions;
+      if (producer_index == minimum_slot_producer_index) {
+         // this is the producer for the minimum slot, go with that
+         return minimum_slot;
       } else {
-         return hard_deadline;
+         // calculate how many rounds are between the minimum producer and the producer in question
+         size_t producer_distance = producer_index - minimum_slot_producer_index;
```

Comment:
Is it guaranteed that `producer_index >= minimum_slot_producer_index`

Reply:
Apparently not, the next line checks:
```
         // check for unsigned underflow
         if (producer_distance > producer_index) {
            producer_distance += active_schedule.size();
         }
```

Reply:
When there is an underflow, I'm not understanding how `producer_distance += active_schedule.size();` addresses the issue?